### DISCUSSION TITLE
What is the best way to put Layer Lock on a mod-tap?

### DISCUSSION THREAD

**Discussion by Dimagog** [getreuer/qmk-keymap#52] [Q&A] [answered]
Hi,
I want to use my left `Ctrl` as a LayerLock key when tapped.
How do I do it right given `LLOCK` cannot be used with `MT()` directly?
I’ve managed to get it to work in a convoluted way, but I’m wondering what would be a good way to do it.

Here is how I did it for now:
1. Made the key for `Ctrl/LLOCK` to be `MT(MOD_LCTL, KC_F17)`
2. Added unused key `NOT_USED`
3. Modified the code in `process_record_user` to call LayerLock like this:
```C
    if (!process_layer_lock(keycode, record, record->tap.count > 0 ? MT(MOD_LCTL, KC_F17) : NOT_USED)) {
        return false;
    }
```

The `:?` conditional is to prevent layer lock activation on **holding** `Ctrl`.

Another option I've considered was to process `MT(MOD_LCTL, KC_F17)` myself and to recursively call `process_record` with a synthetic record for `LLOCK`, but I'm not sure such re-enterancy is supported/encouraged/tolerated by QMK. And there is no function in `layer_lock.c` to activate the lock.

**Comment by getreuer** [comment, marked as answer]
Thank you for the great question! You're right, it's natural to want to combine Layer Lock with some other function on hold.

That's a slick workaround you made with that conditional arg `record->tap.count > 0 ? MT(MOD_LCTL, KC_F17) : NOT_USED`. This makes sense. If you like, you could use the existing `KC_NO` keycode for the purpose of `NOT_USED`.

> Another option I've considered was to process MT(MOD_LCTL, KC_F17) myself and to recursively call process_record with a synthetic record for LLOCK, but I'm not sure such re-enterancy is supported/encouraged/tolerated by QMK.

Yes, I've struggled a fair bit with this approach in Achordion and Repeat Key and have some comments about it. Recursively calling `process_record()` does work, and the Dynamic Macros and Repeat Key core features make use of this internally. When doing this, you need to ensure your logic is such that an infinite recursion is impossible, since getting caught in an infinite recursion would make the keyboard inoperable. The other caveat is when recursing, feature handlers that run before `process_record_user()` will process the event twice, which may confuse things. So yeah, it's dicey and worth avoiding when there's an alternative.

Generally, for combining non-basic keys with MT, there is a pattern suggested in ["changing the tap function"](https://docs.qmk.fm/mod_tap#changing-tap-function) in the mod-tap documentation. This appears to be QMK's recommended solution to the problem. For the case of Layer Lock, the handler should, on tap, call `layer_lock_invert(get_highest_layer(layer_state))` to toggle the lock on the highest active layer. I've added [a section describing this](https://getreuer.info/posts/keyboards/layer-lock/index.html#combine-layer-lock-with-a-mod-tap) to the Layer Lock documentation.
